### PR TITLE
OSDOCS-1810: Add MTU configuration change for installed clusters

### PR DESCRIPTION
**Pull request by jboxman-rh**
- https://issues.redhat.com/browse/OSDOCS-1810

Preview:
- [Changing the MTU for the cluster network](https://deploy-preview-38891--osdocs.netlify.app/openshift-enterprise/latest/networking/changing-cluster-network-mtu.html)

TODO
- Squash commits

### CODE REVIEW

**Comment by jboxman-rh** [comment]
Hi @jcaamano, can you review this docs PR regarding changing the MTU when you have a moment? Thanks!

File: modules/nw-cluster-mtu-change-about.adoc [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+- `spec.migration.mtu.network.from`
+- `spec.migration.mtu.network.to`
+- `spec.migration.mtu.machine.to`
+
+|
+*Cluster Network Operator (CNO)*: Confirms that each field is set to a valid value.
+
+- The `mtu.network.from` field must equal `network.status.clusterNetworkMTU` which is the current MTU of the cluster network.
+- If the `mtu.network.to` field is specified, it must allow for the overlay overhead for the cluster network provider. For OVN-Kubernetes, this value is `100` and for OpenShift SDN this value is `50`.
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change-about.adoc:63]: `mtu.network.to` is mandatory, but nothing prevents it being equal to `mtu.network.from`

File: modules/nw-cluster-mtu-change.adoc [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+$ oc describe network.config cluster
+----
+
+. Get the NetworkManager connection name:
+
+.. To find the primary network interface, enter the following command:
++
+[source,terminal]
+----
+$ oc debug node/<node_name> -- chroot /host ip route list match 0.0.0.0/0 | awk '{print $5 }'
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change.adoc:35]: We are assuming ipv4 throughout the doc. I am not sure if we generally provide guidance for ipv4 only, but ipv6 would be slightly different on the commands (`ip -6 ... <ipv6 address>`). Just a heads up.

File: modules/nw-cluster-mtu-change.adoc [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
++
+[source,terminal]
+----
+$ oc debug node/<node_name> -- chroot /host nmcli c | grep $(ip route list match 0.0.0.0/0 | awk '{print $5 }')
+----
++
+.Example output
+[source,text]
+----
+Wired connection 1  46da4a6a-xxxx-xxxx-xxxx-ac0ca900f213  ethernet  ens3
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change.adoc:54]: This is not used later in the doc for anything?

File: modules/nw-cluster-mtu-change.adoc [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+
+. Prepare your configuration for the hardware MTU:
+
+** If your hardware MTU is specified in a NetworkManager connection configuration, complete the following steps. This approach is the default for {product-title} if you do not explicitly specify your network configuration with DHCP, a kernel command line, or some other method.
+
+... To create a new NetworkManager connection configuration with the hardware MTU, enter the following command:
++
+[source,terminal]
+----
+$ oc debug node/<node_name> -- chroot /host cat /etc/sysconfig/network-scripts/ifcfg-<connection_name> > ifcfg.txt
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change.adoc:65]: This is a bit disjoint to the previous step where we were assuming a NM profile vs here were we are assuming an ifcfg file. They are not the same thing. I don't know which is the default, but I guess the most frequent case would be an auto-generated NM profile.
My guess is that it would be better to doc here how to extract the NM profile obtained on the previous step  and set the MTU there. And then have alternate instructions in case the config is through an ifcfg file if required.

File: modules/nw-cluster-mtu-change.adoc [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+dhcp-option-force=26,<mtu>
+----
++
+--
+where:
+
+`<mtu>`:: Specifies the hardware MTU for the DHCP server to advertise.
+--
+
+** If your hardware MTU is specified with a kernel command line with PXE, update that configuration accordingly.
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change.adoc:186]: Is it actually possible to change the kernel command line after deployment?

Otherwise, NM profiles would be generated from them, and the previous instruction on how to update the NM profile would apply. With the caveat that the `connection.autoconnect-priority` would have to be set to a value above the default of 0.

File: modules/nw-cluster-mtu-change.adoc [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+----
+$ oc patch Network.operator.openshift.io cluster --type=merge --patch \
+  '{"spec": { "migration": { "mtu": { "network": { "from": 1400, "to": 9000 } , "machine": { "to" : 9100} } } } }'
+----
+
+. To confirm that all of the nodes in your cluster were rebooted successfully, enter the following command:
++
+[source,terminal]
+----
+$ oc get nodes
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change.adoc:215]: We would probably want something more similar to steps 4 & 5 of the sdn to ovn-k migration guide to verify that MCO has applied the changes and rebooted.

https://docs.openshift.com/container-platform/4.9/networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.html#nw-ovn-kubernetes-migration_migrate-from-openshift-sdn

For step 5.b, the equivalent to check for would be `ExecStart=/usr/local/bin/mtu-migration.sh`

File: modules/nw-cluster-mtu-change.adoc [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+** If are specifying the new MTU with a NetworkManager connection configuration, enter the following command. The MachineConfig Operator automatically performs a rolling reboot of the nodes in your cluster.
++
+[source,terminal]
+----
+$ for manifest in machine-config-control-plane machine-config-worker; do
+    oc create -f $manifest.yaml
+  done
+----
+
+** If are specifying the new MTU with a DHCP server option or a kernel command line and PXE, reboot each node manually.
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change.adoc:229]: This specific reboot is not strictly necessary, as the reboot on the next step would acquire the changes.

File: modules/nw-cluster-mtu-change.adoc [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+
+. To confirm that all of the nodes in your cluster were rebooted successfully, enter the following command:
++
+[source,terminal]
+----
+$ oc get nodes
+----
+
+. To finalize the MTU migration, enter one of the following commands:
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change.adoc:238]: Add that `The Machine Config Operator performs a rolling reboot of the nodes in the cluster in preparation for the MTU change`

Also, as in the previous comment, we would probably want something more similar to steps 4 & 5 of the sdn to ovn-k migration guide to verify that MCO has applied the changes and rebooted.

In this case, forr step 5.b, the equivalent to check for would be the absence of `ExecStart=/usr/local/bin/mtu-migration.sh`

File: modules/nw-cluster-mtu-change.adoc [hunk 9]
```diff
@@ -0,0 +1,309 @@
+// Module included in the following assemblies:
+//
+// * networking/configuring-cluster-network-mtu.adoc
+
+[id="nw-cluster-mtu-change_{context}"]
+= Changing the cluster MTU
+
+As a cluster administrator, you can change the maximum transmission unit (MTU) for your cluster. The migration is disruptive and nodes in your cluster might be temporarily unavailable as the MTU update rolls out.
+
+.Prerequisites
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change.adoc:10]: I have no idea if it would be safe to assume that all of the cluster nodes are running with the same network configuration. I not, It might be worth it to add a prerequisite on that respect.

File: modules/nw-cluster-mtu-change-about.adoc [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+If the values provided are valid, the CNO writes out a new temporary configuration with the MTU for the cluster network set to the value of the `mtu.network.to` field.
+
+*Machine Config Operator (MCO)*: Performs a rolling reboot of each node in the cluster.
+
+|Reconfigure the MTU of the primary network interface for the nodes on the cluster.
+|N/A
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change-about.adoc:70]: Perhaps we can add a small comment here on the options with which this might be achieved, some of them are:
* Deploying an updated ifcfg or NM profile with the MTU throuh MCO
* Changing the DHCP server MTU setting
* Changing boot parameters (if possible)

File: modules/nw-cluster-mtu-change.adoc [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+. To confirm that all of the nodes in your cluster were rebooted successfully, enter the following command:
++
+[source,terminal]
+----
+$ oc get nodes
+----
+
+. Update the underlying network interface MTU value:
+
+** If are specifying the new MTU with a NetworkManager connection configuration, enter the following command. The MachineConfig Operator automatically performs a rolling reboot of the nodes in your cluster.
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change.adoc:220]: There is an option to pause the machine config pools to apply the MCO manifests and the last step of the migration at the same time and then un-pause after to start the rolling reboot, overall saving one rolling reboot. Example:

Pause the pools
```
  oc patch MachineConfigPool master --type='merge' --patch '{ "spec": { "paused": true } }'
  oc patch MachineConfigPool worker --type='merge' --patch '{ "spec":{ "paused" :true } }'
```

Apply MCO manifests
```
$ for manifest in machine-config-control-plane machine-config-worker; do
    oc create -f $manifest.yaml
  done
```

Apply the last step of the migration
```
$ oc patch Network.operator.openshift.io cluster --type=merge --patch \
  '{"spec": { "migration": null, "defaultNetwork":{ "ovnKubernetesConfig": { "mtu": <mtu> }}}}'
```

Verify both changes took effect on the rendered machine configs.

Unpause the pools
```
  oc patch MachineConfigPool master --type='merge' --patch '{ "spec": { "paused": false } }'
  oc patch MachineConfigPool worker --type='merge' --patch '{ "spec":{ "paused" :false } }'
```

I guess we would go by what QE used to verify, but this could be nice for the customer.

File: modules/nw-cluster-mtu-change.adoc [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+  done
+----
+
+** If are specifying the new MTU with a DHCP server option or a kernel command line and PXE, reboot each node manually.
+
+. To confirm that all of the nodes in your cluster were rebooted successfully, enter the following command:
++
+[source,terminal]
+----
+$ oc get nodes
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change.adoc:235]: Also, as in the previous comment, we would probably want something more similar to steps 4 & 5 of the sdn to ovn-k migration guide to verify that MCO has applied the changes and rebooted.

In this case, forr step 5.b, the equivalent to check for would the files `path: ...` in the rendered configs.

File: modules/nw-cluster-mtu-change.adoc (hunk 2 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:35]: @jcaamano, so that's interesting and I had not considered that; If you have an IPv4+IPv6 cluster, would you do the migration twice? I don't think we support IPv6-only clusters.

File: modules/nw-cluster-mtu-change.adoc (hunk 3 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:54]: @jcaamano, this is the NetworkManager profile name that's used later for a user that wants to use this to set the MTU. But maybe it makes sense to do this elsewhere.

File: modules/nw-cluster-mtu-change.adoc (hunk 4 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:65]: @jcaamano, oh, I pulled this from an unpublished KCS; I don't actually have any idea how to do this, so this was my best guess. Specifically: https://access.redhat.com/solutions/5307301

File: modules/nw-cluster-mtu-change.adoc (hunk 5 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:186]: @jcaamano, interesting; Not sure, I'm open to suggestions about the best way(s) to advise users to modifying the underlying MTU. This is mostly based on an unpublished KCS that I found for how to do this.

File: modules/nw-cluster-mtu-change.adoc (hunk 6 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:215]: @jcaamano, perfect, thanks!

File: modules/nw-cluster-mtu-change.adoc (hunk 2 above)
Comment (jcaamano) [reply modules/nw-cluster-mtu-change.adoc:35]: Ah, I thought we did support IPv6-only clusters as we have CI for it but indeed I did not find it documented anywhere.

File: modules/nw-cluster-mtu-change.adoc (hunk 3 above)
Comment (jcaamano) [reply modules/nw-cluster-mtu-change.adoc:54]: But I don't see how this is used later in the document?

File: modules/nw-cluster-mtu-change.adoc (hunk 4 above)
Comment (jcaamano) [reply modules/nw-cluster-mtu-change.adoc:65]: So for an NM connection profile, taking the UUID from the previous example....

```
# get the profile path on the node
profile="$(oc debug node/<node_name> -- chroot /host nmcli -g UUID,FILENAME c show | grep 46da4a6a-xxxx-xxxx-xxxx-ac0ca900f213 | cut -d: -f2)"

# extract the profile
oc debug node/<node_name> -- chroot /host cat "$profile" > "$(basename $profile)"

# edit profile
# set 802-3-ethernet.mtu for MTU
# set connection.autoconnect-priority to increase priority over other existing profiles for the same device if required
# remove connection.uuid
# change connection.id to rename the connection as required
vi "$(basename $file)"
```

Then you would follow on with the MacineConfig part to write the profile in `/etc/NetworkManager/system-connections/`

Hopefully we could refer to RHEL8 docs i.e.
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_and_managing_networking/assembly_manually-creating-networkmanager-profiles-in-key-file-format_configuring-and-managing-networking

File: modules/nw-cluster-mtu-change.adoc (hunk 5 above)
Comment (jcaamano) [reply modules/nw-cluster-mtu-change.adoc:186]: No idea myself. Will let other chime in on this one.

File: modules/nw-cluster-mtu-change.adoc (hunk 9 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:10]: @jcaamano, can you elaborate a bit on _same network configuration_? Is this the same MTU everywhere or the same network card or the same routes or something else? Thanks!

File: modules/nw-cluster-mtu-change.adoc (hunk 12 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:235]: @jcaamano, what is `path: ...` in this context?

File: modules/nw-cluster-mtu-change.adoc (hunk 3 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:54]: Looks like it doesn't work, at least locally. I lack the skill to determine how to fix it.

```
oc debug node/crc-8rwmc-master-0 -- chroot /host nmcli c | grep $(ip route list match 0.0.0.0/0 | awk '{ print $5 }')
Starting pod/crc-8rwmc-master-0-debug ...
To use host binaries, run `chroot /host`

Removing debug pod ...
```

```
oc debug node/crc-8rwmc-master-0 -- chroot /host ip route list match 0.0.0.0/0 | awk '{ print $5 }'
Starting pod/crc-8rwmc-master-0-debug ...
To use host binaries, run `chroot /host`
enp1s0

Removing debug pod ...
```

```
oc debug node/crc-8rwmc-master-0 -- chroot /host nmcli c | grep $(echo 'enp1s0')
Starting pod/crc-8rwmc-master-0-debug ...
To use host binaries, run `chroot /host`
Wired connection 1  3e96a02b-88e0-3707-ad5d-61db28678130  ethernet  enp1s0

Removing debug pod ...
```

But somehow this is possible.

File: modules/nw-cluster-mtu-change.adoc (hunk 12 above)
Comment (jcaamano) [reply modules/nw-cluster-mtu-change.adoc]: Lookup in the rendered machine config for the file with the same path as that provided in the machine config that was constructed on this step. So basically grepping for the same path. Taken from step 5.b:

```
oc get machineconfig <config_name> -o yaml | grep path: <path>
```

File: modules/nw-cluster-mtu-change.adoc (hunk 3 above)
Comment (jcaamano) [reply modules/nw-cluster-mtu-change.adoc:78]: I am confused about where that command comes from, but yeah, that won't work. Divide it in two:

```
dev=$(oc debug node/crc-8rwmc-master-0 -- ip route show default | awk '{ print $5; exit }')
profile=$(oc debug node/crc-8rwmc-master-0 -- chroot /host nmcli -g DEVICE,CON-UUID d | grep $dev | cut -d: -f2)
```

File: modules/nw-cluster-mtu-change-about.adoc (hunk 10 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change-about.adoc:70]: Added.

File: modules/nw-cluster-mtu-change.adoc (hunk 9 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:10]: @jcaamano, I moved this into the procedure:

> If your hardware MTU is specified in a NetworkManager connection configuration, complete the following steps. This approach is the default for {product-title} if you do not explicitly specify your network configuration with DHCP, a kernel command line, or some other method. Your cluster nodes must all use the same underlying network configuration for the following procedure to work unmodified.

File: modules/nw-cluster-mtu-change.adoc (hunk 3 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:77]: @jcaamano, I did something similar and moved the step; It only seems applicable for the MachineConfig approach.

File: modules/nw-cluster-mtu-change.adoc (hunk 9 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:10]: @jcaamano, so I think in theory, someone could setup additional MachineConfigPools and have different MTU configurations, but I wanted to avoid covering every possible scenario for this. If someone from QE has ideas for DHCP or other approaches, I can add them, but I can't derive them locally as I lack the expertise and hardware setup to do so. I'm trying to cover what I hope are the most common cases.

File: modules/nw-cluster-mtu-change.adoc (hunk 8 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:411]: @jcaamano, works for me; I knew that validation procedure would pay dividends someday!

**Comment by jboxman-rh** [comment]
@jcaamano, I've gone through and completed the suggested changes, thanks!

**Comment by jboxman-rh** [comment]
Hi @brahaney, can you take a look at this draft documentation for changing the cluster MTU? Thanks!

File: modules/nw-cluster-mtu-change.adoc [hunk 13]
```diff
@@ -0,0 +1,482 @@
...
+
+. Prepare your configuration for the hardware MTU:
+
+** If your hardware MTU is specified in a NetworkManager connection configuration, complete the following steps. This approach is the default for {product-title} if you do not explicitly specify your network configuration with DHCP, a kernel command line, or some other method. Your cluster nodes must all use the same underlying network configuration for the following procedure to work unmodified.
+
+... To find the primary network interface, enter the following command:
++
+[source,terminal]
+----
+$ oc debug node/<node_name> -- chroot /host ip route list match 0.0.0.0/0 | awk '{print $5 }'
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change.adoc:52]: I just noticed that for ovn-k, this would give you br-ex as interface name and that's not what we want. So this step instruction would only be valid for SDN.

For OVN-K, the interface we want is the system interface slaved on br-ex, several ways to get it:

`oc debug node/<node-name> -- chroot /host nmcli -g connection.interface-name c show ovs-if-phys0`

or

`oc debug node/<node-name> -- chroot /host ovs-vsctl --columns name find interface type=system | cut -d: -f2`

File: modules/nw-cluster-mtu-change.adoc [hunk 14]
```diff
@@ -0,0 +1,482 @@
...
+where:
+
+`<node_name>`:: Specifies the name of a node in your cluster.
+--
+
+... To find the connection profile that NetworkManager created for the interface name returned from the previous command, enter the following command:
++
+[source,terminal]
+----
+$ oc debug node/<node_name> -- chroot /host nmcli c | grep <interface>
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change.adoc:65]: For OVN-K, once you obtain the appropriate interface name as per my last comment above, this output will either:

* give three profiles, one is the one the node originally had if it had one configured, and the other two are the ones that we internally generate from the former, named `ovs-if-phys0` and `ovs-port-phys0`. The appropriate to use would be the the one the node originally had.
* give only two profiles that we internally generate named `ovs-if-phys0` and `ovs-port-phys0`. In this case, the node was never originally configured with a profile and one ephemeral default DHCP profile was once generated that might not longer be there. In this case, the customer has to use a default DHCP profile as a starting point.

Example output of the last case
```
❯ oc debug node/jcaamano-mtu-migratio-zxct2-master-0 -- chroot /host nmcli c | grep ens4                                   
Starting pod/jcaamano-mtu-migratio-zxct2-master-0-debug ...
To use host binaries, run `chroot /host`
ovs-if-phys0        353774d3-0d3d-4ada-b14e-cd4d8824e2a8  ethernet       ens4   
ovs-port-phys0      332ef950-b2e5-4991-a0dc-3158977c35ca  ovs-port       ens4  
```

Given the complexity, would we probably better suggest as an example in general to build a default DHCP profile with the MTU setting for a default (IPI) case, and mention for a non-default case (UPI) that the customer should customize the original deployed profile, which I guess should be aware of?

File: modules/nw-cluster-mtu-change.adoc [hunk 15]
```diff
@@ -0,0 +1,482 @@
...
+nm-generated=true
+----
+
+... Edit the NetworkManager configuration file saved in the `config.nmconnection` file from the previous step:
++
+--
+**** Set the following values:
+***** `802-3-ethernet.mtu`: Specify the MTU for the primary network interface of the system.
+***** `connection.interface-name`: Optional: Specify the network interface name that this configuration applies to.
+***** `connection.autoconnect-priority`: Optional: Specify an integer priority value greater than any other ConnectionManager connection profiles to ensure this profile is used.
```
Comment (jcaamano) [inline modules/nw-cluster-mtu-change.adoc:150]: I would rephrase to `Consider specifying an integer priority value above 0 to ensure this profile is used over other profiles for the same interface`. For OVN-K, priority should not be set at or above 100.

File: modules/nw-cluster-mtu-change.adoc (hunk 14 above)
Comment (jcaamano) [reply modules/nw-cluster-mtu-change.adoc:65]: Sorry for not noticing this before and all the extra work I might have caused.

**Review by jcaamano — CHANGES_REQUESTED** [review]

File: modules/nw-cluster-mtu-change-about.adoc [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+- Different endpoints in the cluster might use different MTUs during the migration. As a result it is possible that some network traffic is dropped until the MTU change is finalized.
+
+[id="mtu-value-selection_{context}"]
+== MTU value selection
+
+When planning your MTU migration there are two related but distinct MTU values to consider.
+
+* *Hardware MTU*: This MTU value is set based on the specifics of your network infrastructure.
+* *Cluster network MTU*: This MTU value is always less than your hardware MTU to account for the cluster network overlay overhead. The specific overhead is determined by your cluster network provider:
+** *OVN-Kubernetes*: `100` bytes
```
Comment (kyrtapz) [inline modules/nw-cluster-mtu-change-about.adoc:35]: I think it would be good to add a statement that only `OVN-Kubernetes` and `OpenShift SDN` are supported for MTU migration

File: modules/nw-cluster-mtu-change-about.adoc [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+
+When you initiate an MTU change on your cluster the following effects might impact service availability:
+
+- At least two rolling reboots are required to complete the migration to a new MTU. During this time, some nodes are not available as they restart.
+
+// https://github.com/openshift/enhancements/pull/926/files#diff-166e452788d3711a8dfa7254d867f156c7b7c659f172e633acbf6852504ae358R180
+- Specific applications deployed to the cluster with shorter timeout intervals than the absolute TCP timeout interval might experience disruption during the change.
+
+// https://github.com/openshift/enhancements/pull/926/files#diff-166e452788d3711a8dfa7254d867f156c7b7c659f172e633acbf6852504ae358R190
+- Different endpoints in the cluster might use different MTUs during the migration. As a result it is possible that some network traffic is dropped until the MTU change is finalized.
```
[the remark this answers is not in the document]
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change-about.adoc:26]: I'll omit the third bullet, then, which seems covered by the first.

File: modules/nw-cluster-mtu-change-about.adoc [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+- `spec.migration.mtu.network.from`
+- `spec.migration.mtu.network.to`
+- `spec.migration.mtu.machine.to`
+
+|
+*Cluster Network Operator (CNO)*: Confirms that each field is set to a valid value.
+
+- The `mtu.network.from` field must equal `network.status.clusterNetworkMTU` which is the current MTU of the cluster network.
+- The `mtu.network.to` field must allow for the overlay overhead for the cluster network provider. For OVN-Kubernetes, this value is `100` and for OpenShift SDN this value is `50`. The value of `mtu.network.to` can equal the value of `mtu.network.from`.
+
```
[the remark this answers is not in the document]
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change-about.adoc:64]: @jcaamano, I thought it was `network.{from,to}` that reflects the value for the cluster network provider? And that this value needs to accommodate the SDN overhead? So that this is set relative to the network link MTU?

File: modules/nw-cluster-mtu-change.adoc [hunk 19]
```diff
@@ -0,0 +1,482 @@
...
+
+As a cluster administrator, you can change the maximum transmission unit (MTU) for your cluster. The migration is disruptive and nodes in your cluster might be temporarily unavailable as the MTU update rolls out.
+
+.Prerequisites
+
+* You installed the OpenShift CLI (`oc`).
+* You are logged in to the cluster with a user with `cluster-admin` privileges.
+* You identified the target MTU for your cluster. The correct MTU varies depending on the cluster network provider that your cluster uses:
+** *OVN-Kubernetes*: The cluster MTU must be set to `100` less than the lowest MTU value in your cluster.
+** *OpenShift SDN*: The cluster MTU must be set to `50` less than the lowest MTU value in your cluster.
```
[the remark this answers is not in the document]
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:16]: @jcaamano, interesting; That's the oppose of how I've seen it explained.

In this context, _cluster network_ refers to the SDN or the pod network overlay, rather than the underlying physical network adapter on the nodes themselves.

But maybe I have it backwards?

Thanks!

File: modules/nw-cluster-mtu-change.adoc (hunk 13 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:52]: @jcaamano, awesome, thank you!

File: modules/nw-cluster-mtu-change.adoc (hunk 14 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:65]: @jcaamano, that sounds unfortunately complicated. I'm not sure who might know how to accomplish this via DHCP, as I don't have any experience in the installation area. Maybe I can find someone, but probably not before docs freeze given the complexity in all of this.

Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:65]: Looks like we have this for user-provisioned: https://docs.openshift.com/container-platform/4.9/installing/installing_bare_metal/installing-bare-metal.html#installation-user-infra-machines-advanced_network_installing-bare-metal

File: modules/nw-cluster-mtu-change.adoc (hunk 8 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:411]: @jcaamano, added that text back in.

File: modules/nw-cluster-mtu-change-about.adoc (hunk 18 above)
[the remark this answers is not in the document]
Comment (jcaamano) [reply modules/nw-cluster-mtu-change-about.adoc:63]: You are right. I was getting confused because we are mixing the term `cluster network` and `cluster network provider`. Previously in the document you define `Cluster network MTU` and `Hardware MTU` so we should probably stick to those terms to avoid confusion.

I think this does not make sense
`The value of mtu.network.to can equal the value of mtu.network.from`
Although technically correct, it does not make a lot of sense if the purpose of the instruction is actually to change the cluster network MTU. I did give a comment about `mtu.machine.to` in this regard before but not about ` mtu.network.to`

Overall, I might rephrase to something more like:
```
- The `mtu.machine.to` must be set to the a desired Hardware MTU or to the current Hardware MTU. Note that this value is only informative to the MTU migration procedure and that the Hardware MTU has to be configured by other means. Some suggestion are provided later on.
- The `mtu.network.from` field must equal `network.status.clusterNetworkMTU` which is the current MTU of the cluster network.
- The `mtu.network.to` field must be set to the new desired cluster network MTU and must be lower than the Hardware MTU to allow for the overlay overhead of the cluster network provider. For OVN-Kubernetes, the overhead is `100` and for OpenShift SDN this overhead is `50`. 
```

File: modules/nw-cluster-mtu-change.adoc (hunk 14 above)
Comment (jcaamano) [reply modules/nw-cluster-mtu-change.adoc:97]: A default DHCP profile might look like this

```
[connection]
id=<id>
type=ethernet
autoconnect-priority=<0-99>
interface-name=<iface>

[ethernet]
mtu=<mtu>

[ipv4]
method=auto

[ipv6]
addr-gen-mode=stable-privacy
method=auto
```

But we can always refer to the appropriate RHEL NM documentation, for example:

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/networking_guide/sec-configuring_ip_networking_with_nmcli#sec-Adding_and_Configuring_a_Dynamic_Ethernet_Connection_with_nmcli

**Comment by jboxman-rh** [comment]
@brahaney, any news? Thanks!

File: modules/nw-cluster-mtu-change.adoc (hunk 14 above)
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change.adoc:97]: @jcaamano, I'll include this, thanks!

**Review by brahaney — APPROVED** [review]
Looks good to me

File: modules/nw-cluster-mtu-change.adoc [hunk 20]
```diff
@@ -0,0 +1,512 @@
...
+[connection]
+id=Primary network interface
+type=ethernet
+autoconnect-priority=10
+interface-name=enp1s0
+[802-3-ethernet]
+mtu=8051
+----
+
+... To base64 encode the NetworkManager configuration for inclusion in a `MachineConfig` object, enter the following command:
```
Comment (sagidlow) [inline modules/nw-cluster-mtu-change.adoc:214]: Should base64 be in backticks here? Not sure, but just a thought.

File: networking/changing-cluster-network-mtu.adoc [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+
+toc::[]
+
+[role="_abstract"]
+As a cluster administrator, you can change the MTU for the cluster network after cluster installation. This change is disruptive as cluster nodes must be rebooted to finalize the MTU change. You can change the MTU only for clusters using the OVN-Kubernetes or OpenShift SDN cluster network providers.
+
+include::modules/nw-cluster-mtu-change-about.adoc[leveloffset=+1]
+include::modules/nw-cluster-mtu-change.adoc[leveloffset=+1]
+
+[id="{context}-additional-resources"]
```
Comment (sagidlow) [inline networking/changing-cluster-network-mtu.adoc:15]: Above this line, we need the following: `[role="_additional-resources"]` per the [Anchoring "Prerequisites", "Additional resources", and "Next steps" titles in assemblies](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-prerequisites-additional-resources-and-next-steps-titles-in-assemblies)

File: modules/nw-cluster-mtu-change.adoc [hunk 22]
```diff
@@ -0,0 +1,512 @@
...
+where:
+
+`<mtu>`:: Specifies the new cluster network MTU that you specified with `<overlay_to>`.
+--
+
+.Verification
+
+You can verify that a node in your cluster uses an MTU that you specified in the previous procedure.
+
+. To get the current MTU for the cluster network, enter the following command:
```
Comment (sagidlow) [inline modules/nw-cluster-mtu-change.adoc:478]:
```suggestion
. To retrieve the current MTU for the cluster network, enter the following command:
```
Suggestion only, retrieve seemed like a better word here.

File: networking/changing-cluster-network-mtu.adoc [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+== Additional resources
+
+* xref:../installing/installing_bare_metal/installing-bare-metal.adoc#installation-user-infra-machines-advanced_network_installing-bare-metal[Using advanced networking options for PXE and ISO installations]
+* link:https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_and_managing_networking/assembly_manually-creating-networkmanager-profiles-in-key-file-format_configuring-and-managing-networking[Manually creating NetworkManager profiles in key file format]
```
Comment (sagidlow) [inline networking/changing-cluster-network-mtu.adoc:19]: My only concern with links on Line 19 and Line 20 is that we take our customers out of their doc experience and put them on the portal. This isn't so bad for customers viewing their docs on the portal but if they are viewing it on docs.openshift.com they're being redirected to a different user experience. Is there a way to get to this content another way? If not, then ignore this comment.

**Review by sagidlow — COMMENTED** [review]
Looks great, just a few nit comments. Let me know if you have any questions.

**Comment by jboxman-rh** [comment]
@sagidlow thank you!

File: networking/changing-cluster-network-mtu.adoc (hunk 23 above)
Comment (jboxman-rh) [reply networking/changing-cluster-network-mtu.adoc:19]: @sagidlow, true, but the RHEL docs are available only on the customer portal.

File: modules/nw-cluster-mtu-change-about.adoc (hunk 18 above)
[the remark this answers is not in the document]
Comment (jcaamano) [reply modules/nw-cluster-mtu-change-about.adoc:63]: I still miss something here that at least conveys the idea that `mtu.machine.to` must be set to the Hardware MTU in a similar way as I phrased above.

[the remark this answers is not in the document]
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change-about.adoc:63]: @jcaamano, I'll look, thanks!

File: networking/changing-cluster-network-mtu.adoc (hunk 23 above)
Comment (sagidlow) [reply networking/changing-cluster-network-mtu.adoc:20]: @jboxman-rh that's what I figured but thought I would voice my concern anyway :)

File: modules/nw-cluster-mtu-change-about.adoc (hunk 18 above)
[the remark this answers is not in the document]
Comment (jboxman-rh) [reply modules/nw-cluster-mtu-change-about.adoc]: @jcaamano, updated, can you confirm? Thanks!
